### ISSUE TITLE
The application consumes almost 100% CPU (1 thread) all the time.

### ISSUE THREAD

**Issue by vidcentum** [Stiffstream/sobjectizer#18]
Hi,
I am using SObjectizer in our projects. I have some observation on the CPU utilisation. The following is a simple code with one thread handling messages on 2 channels. 
// Channel 1   -> process a request and post a message to channel 2
// Channel 2   -> process a request and post a message to channel 1 
And the loop continues.

**Observation**: The application consumes almost 100% CPU (1 thread) all the time. 

**Query**: Please help make it CPU efficient / event based? I admit I am still learning the SObjectizer library.

**Sample Code:**

```
#include <thread>
#include <memory>
#include <random>

// SObjectizer is one of a few cross-platform and OpenSource "actor frameworks" for C++. 
// But SObjectizer supports not only Actor Model, but also Publish-Subscribe Model and CSP-like channels. 
// ver 5.7
#include <so_5/all.hpp>

// Message structure
struct poll_device_events_request_t { };

void thread_func(so_5::mchain_t primary_channel, so_5::mchain_t secondary_channel)
{
  // The stuff necessary for random pause generation.
  std::random_device rd;
	std::mt19937 generator{rd()};
	std::uniform_int_distribution<> pause_generator{5000, 10000}; // Delay between 5 - 10 seconds.
  
  // This flag will be set to 'true' when some of channels will be closed.
	bool stop = false;
  auto prepared = so_5::prepare_select(
    so_5::from_all().handle_all()
      // If some channel gets closed we should set out 'stop' flag.
      .on_close([&stop](const auto &) { stop = true; })
      // A predicate for stopping select() function.
      .stop_on([&stop]{ return stop; }),

    // Read and handle poll_device_events_request messages from channel_.
    so_5::receive_case(primary_channel, 
      [&](poll_device_events_request_t){
        
        std::cout << "Received message on primary channel..." << std::endl;
        // Do some work.
        
        // Generate a random pause.
        const std::chrono::milliseconds pause{pause_generator(generator)};
        
        // Send a message to secondary channel - delayed message.
        so_5::send_delayed<poll_device_events_request_t>(secondary_channel, pause);
      }
    ),
    so_5::receive_case(secondary_channel, 
      [&](poll_device_events_request_t){
        std::cout << "Received message on secondary channel..." << std::endl;
        // Do some work.
        
        // Generate a random pause for processing of that request.
        const std::chrono::milliseconds pause{pause_generator(generator)};
        
        // Delay processing of that request by using delayed message.
        so_5::send_delayed<poll_device_events_request_t>(primary_channel, pause);
      }
    )
  );
  
  while (!stop) {
    so_5::select( prepared );
  }
}

// Let's create two channels - 
// Channel 1 -> process a request and post a message to channel 2
// Channel 2 -> process a request and post a message to channel 1
  
int main() {
  std::shared_ptr<so_5::wrapped_env_t> s_node_env_obj_ = std::make_shared<so_5::wrapped_env_t>();
  
  auto primary_channel_ = so_5::create_mchain(*s_node_env_obj_);
  auto primary_channel_closer = so_5::auto_close_drop_content(primary_channel_);
	
  auto secondary_channel_ = so_5::create_mchain(*s_node_env_obj_);
  auto secondary_channel_closer = so_5::auto_close_drop_content(secondary_channel_);
  
  // Thread objects for workers.
	std::thread nthread_ {thread_func, primary_channel_, secondary_channel_};
	auto nthread_joiner = so_5::auto_join( nthread_ );

  // Initiate the first message.
  so_5::send<poll_device_events_request_t>(primary_channel_);

  nthread_.join();
  return 0;
}
```

**Comment by eao197** [comment]
Thanks for reporting this. I'll investigate what is going on here.

**Comment by eao197** [comment]
It looks like a flaw in the implementation of the new version of select. There are too many internal notifications about readiness of `primary_channel` and `secondary_channel` for accepting new outgoing messages. But because there is no `send_case`s those notifications are ignored inside of `select`. However, the handling of those notifications eats CPU.

I'll take the time to fix this issue.

**Comment by vidcentum** [comment]
@eao197 Thank you for your quick response. I observe the similar CPU utilisation in Restinio as well. I will test once more and submit the observation.

Best Regards
Maruthi

**Comment by eao197** [comment]
As a workaround, you can add a call to `empty_timeout` during the preparation for select:
```cpp
  auto prepared = so_5::prepare_select(
    so_5::from_all().handle_all()
      .empty_timeout(std::chrono::hours(1))
      // If some channel gets closed we should set out 'stop' flag.
      .on_close([&stop](const auto &) { stop = true; })
```
It seems that [`std::condition_variable::wait_for`](https://en.cppreference.com/w/cpp/thread/condition_variable/wait_for) doesn't handle big values of `rel_time` correctly. Calling of `empty_timeout` with smaller value leads to passing the smaller value to `condition_variable::wait_for` and `wait_for` doesn't return immediately. In that case, CPU consumption dropped to zero because there is no many internal notifications inside `select`.

**Comment by vidcentum** [comment]
@eao197 Thank you once again for a quick fix (workaround). The workaround drops the CPU consumption. Also, I applied this patch for Restinio (without .handle_all()) and observed the CPU consumption dropped well.
Thank you. 

Best Regards
Maruthi

**Comment by eao197** [comment]
The fix is now in '5.7-dev-5.7.2' development branch. And will be a part of the upcoming v.5.7.2. But I can't predict now when this version will be released.

**Comment by vidcentum** [comment]
Thank you, @eao197 Will take the latest dev branch. I will go ahead and close the issue for now.
